### PR TITLE
mrc-2492 Implement summary table for strategising across regions

### PR DESCRIPTION
**Pull request by mwoodbri**
Changes:
- Add strategise page: budget input form and summary table ([mockups](https://imperiallondon.sharepoint.com/:w:/r/sites/MRCGIDA-mint-dev/Shared%20Documents/mint-dev/Optimisation/Stategize%20across%20regions%20-%20mockups.docx?d=w989d5cc08537432287ec7ebe14a912b2&csf=1&web=1)). Details charts/table to follow in mrc-2493 and mrc-2494.
- Apply limit of 15 regions to place an upper bound on time taken in optimiser
- Add mutation/action to clear API errors - previously these just accumulated
- Remove vestigial(?) `package-lock.json`

To test:
- Visit localhost:8080/?stratAcrossRegions=true
- Add at least two regions, each of which has at least one intervention configured i.e. >0% ITN and/or IRS (the optimiser will work with no interventions but will return very uninteresting results)
- Click on "Strategize across regions" in the toolbar

Notes:
- There is definitely scope to turn the strategies table into a more metadata-driven component. In particular the display options  parallel some of the dynamic chart configuration we use in the rest of the app. This wouldn't be trivial though, especially dealing with a dynamic number of columns. At the same time it might make sense to also turn use vue-dynamic-form - especially if we were to add an inline (i.e. horizontal layout). Keen to hear thoughts on this - would have been hard to do speculatively but may merit a follow-up ticket.
- There is now something of a disconnect between the per-region part of the app and the new page which operates across regions. In particular the global budget needs to be accessible to each region because it is used in a formula in the graph metadata. This necessitates copying the budget into each region's settings when it changes. This also means that there is part of the settings that isn't derived directly from the dynamic forms used to otherwise configure each region. I can't see a way to resolve this without a more significant restructure of the app. If the dynamic form allowed hidden fields then we could somehow link one with a reference to a setting elsewhere in the store/state, but that's not ideal.
- There are a couple of usability issues remaining (now noted on top-level ticket mrc-2434):
  - There has never been a project page, only regions. So to strategise for a project you first have to select a region, which enables the strategise tab if the implicitly-selected project has more than one region
  - If you do create many regions (e.g. >5) then the strategies table can get very crowded
- Feature flag will be removed in mrc-2505
- This branch depends on mrc-ide/mintr#59

To-do:
- [x] Revert to mintr master

### CODE REVIEW

File: src/app/static/src/app/components/strategisePage.vue [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    </div>
+    <div class="form mt-5">
+      <b-form inline class="justify-content-center" @submit.prevent="submit">
+        <label class="mr-sm-2" for="inline-form-input-budget">Total available budget</label>
+        <span class="icon-small mr-sm-3"
+              v-tooltip="'The total available budget to cover the next 3-years is required to assess the most feasible intervention options across regions'">
+          <help-circle-icon></help-circle-icon>
+        </span>
+        <b-input-group prepend="$USD" class="mb-2 mr-sm-3 mb-sm-0">
+          <b-form-input id="inline-form-input-budget" v-model="currentProject.budget" type="number" min="0"
```
Comment (mwoodbri) [inline src/app/static/src/app/components/strategisePage.vue:31]: The `v-model` here immediately propagates changes in global budget to the project, and therefore to each region via `RootMutation.SetCurrentRegionInterventionSettings`. The alternative is to only do this when the form is submitted - I'm not sure which is preferable.

File: src/app/static/src/app/components/strategisePage.vue [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+      return this.showDocumentation ? "chevron-up-icon" : "chevron-down-icon";
+    }
+  },
+  methods: {
+    ...mapActions({
+      strategise: RootAction.Strategise,
+      dismissErrors: RootAction.DismissErrors
+    }),
+    submit: async function () {
+      this.strategising = true;
```
Comment (mwoodbri) [inline src/app/static/src/app/components/strategisePage.vue:120]: This could be swapped for a flag in the global state, but this seems unnecessary given that the only consumer is this component. Opinions welcome.

File: src/app/static/src/app/components/figures/strategise/strategiesTable.vue [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+const names: Record<string, string> = {
+  "irs": "IRS only",
+  "llin-pbo": "Pyrethroid-PBO ITN only",
+  "irs-llin-pbo": "Pyrethroid-PBO ITN with IRS*",
+  "llin": "Pyrethroid LLIN only",
+  "irs-llin": "Pyrethroid LLIN with IRS*",
+  "none": "No intervention"
+};
+
+const colours: Record<string, string> = {
```
Comment (EmmaLRussell) [inline src/app/static/src/app/components/figures/strategise/strategiesTable.vue:19]: Nice use of the bootstrap classes, but it does look a bit eccentric! I'd vote for using a colour palette that matches those used in the charts a little more closely and creating our own css classes with those. Hm, I see you've created classes below, are 'primary' etc the only options the BTable will accept in cellVariants? Maybe a comment to this effect in that case. 

More generally, I guess colours and intervention labels are the sort of metadata values which is would make sense to fetch from mintr, since those values are already coming from mintr for the regions page. But that can be a future modification if desirable.

File: src/app/static/src/app/components/figures/strategise/strategiesTable.vue [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+<script lang="ts">
+import {BTable} from "bootstrap-vue";
+import Vue from "vue";
+import {StrategyWithThreshold} from "../../../models/project";
+
+const names: Record<string, string> = {
+  "irs": "IRS only",
+  "llin-pbo": "Pyrethroid-PBO ITN only",
+  "irs-llin-pbo": "Pyrethroid-PBO ITN with IRS*",
```
Comment (EmmaLRussell) [inline src/app/static/src/app/components/figures/strategise/strategiesTable.vue:13]: It seems a bit odd to have these values asterisked on a page which doesn't include the help text for them. Is there a plan to add that into the Strategize page?

File: src/app/static/src/app/components/strategisePage.vue (hunk 2 above)
Comment (EmmaLRussell) [reply src/app/static/src/app/components/strategisePage.vue:120]: Don't mind, happy with this way.

File: src/app/static/src/app/localStorageManager.ts [hunk 5]
```diff
@@ -14,7 +14,8 @@ const stripDataFromProject = (project: Project): Project => {
     return {
         ...project,
         currentRegion: stripDataFromRegion(project.currentRegion),
-        regions: project.regions.map(stripDataFromRegion)
+        regions: project.regions.map(stripDataFromRegion),
+        strategies: []
```
Comment (EmmaLRussell) [inline src/app/static/src/app/localStorageManager.ts:18]: Should strategies go into `serialiseState` as well to persist them?

**Review by EmmaLRussell — COMMENTED** [review]
This looks great, and I can certainly understand taking the approach, it would have been very difficult using the pure metadata approach. 

>In particular the display options parallel some of the dynamic chart configuration we use in the rest of the app. This wouldn't be trivial though, especially dealing with a dynamic number of columns

Agreed, I can see how you could sort of half-drive it from the backend but I'm not sure it would be worth it. 

>At the same time it might make sense to also turn use vue-dynamic-form - especially if we were to add an inline (i.e. horizontal layout)

Not sure I understand what you mean by this - dynamic form for the table layout? So, treat it like a read-only form? Again, the metadata would need to be able to deal with the variable number of regions, so that sounds pretty tricky. 

> I can't see a way to resolve this without a more significant restructure of the app. If the dynamic form allowed hidden fields then we could somehow link one with a reference to a setting elsewhere in the store/state, but that's not ideal.

As I said, I'd imagined having 'project level settings' which could be accessible to the graphing config in the same way as it looks in both baseline and intervention settings for values. Those wouldn't need to be derived from options config, they could just be settings the project itself initialises (and initially would just contain budgetAllZones). 

Alternatively - hidden dynamic form values did occur to me. But like you say, not much value in that unless it can be linked to the value in the state, which is another whole level of stuff. i think I'd favour project settings. 

I ended up with a couple of duplicate strategies, is that supposed to be possible? It's because that's the best strategy at both those percentages? Could be confusing to the user? Not sure what best way of indicating it would be.  
Screenshot from 2021-07-22 11-03-48

Edge case, but I think I get this error if I've created a region but not changed any settings for it yet:
Screenshot from 2021-07-22 11-53-05

It did mean I got to test the nice dismissible errors!

File: src/app/static/src/app/components/figures/strategise/strategiesTable.vue (hunk 4 above)
Comment (mwoodbri) [reply src/app/static/src/app/components/figures/strategise/strategiesTable.vue:13]: Good point - will add to footer of "How to use this page", consistent with regions page

File: src/app/static/src/app/localStorageManager.ts (hunk 5 above)
Comment (mwoodbri) [reply src/app/static/src/app/localStorageManager.ts:18]: Stripped these deliberately as they're not user input - I can't think of a good reason to serialise them. They use up localStorage, don't take long to regenerate and if anything I rather encourage re-generation in case we identify any bugs in the optimiser.

File: src/app/static/src/app/components/figures/strategise/strategiesTable.vue (hunk 3 above)
Comment (mwoodbri) [reply src/app/static/src/app/components/figures/strategise/strategiesTable.vue:19]: Yep - colour variants are an expectation of BTable. Will add explanatory comment - thanks.

File: src/app/static/src/app/components/strategisePage.vue (hunk 1 above)
Comment (mwoodbri) [reply src/app/static/src/app/components/strategisePage.vue:31]: > I think this does make most sense in terms of when it updates - other settings take effect immediately. I'm slightly surprised that works though, since you're not updating the region's setting by setting that value in the project - is it because the form always emits values when it loads, and gets those updated from the current project budget at that point?

Exactly

> We've generally steered clear of directly modifying state values through `v-model`, would usually make use of a mutation instead. That mutation could potentially do the updating of the region's settings, though apparently it's not needed!

Mutation is a good idea - not least because it would make it easier to subsequently change the behaviour by keeping the same mutation but trigger it on form submission itself. Will add.

**Comment by mwoodbri** [comment]
> > At the same time it might make sense to also turn use vue-dynamic-form - especially if we were to add an inline (i.e. horizontal layout)
> 
> Not sure I understand what you mean by this - dynamic form for the table layout? So, treat it like a read-only form? Again, the metadata would need to be able to deal with the variable number of regions, so that sounds pretty tricky.

I meant the budget input form itself - it currently looks a bit like a dynamic form but it isn't!

> > I can't see a way to resolve this without a more significant restructure of the app. If the dynamic form allowed hidden fields then we could somehow link one with a reference to a setting elsewhere in the store/state, but that's not ideal.
> 
> As I said, I'd imagined having 'project level settings' which could be accessible to the graphing config in the same way as it looks in both baseline and intervention settings for values. Those wouldn't need to be derived from options config, they could just be settings the project itself initialises (and initially would just contain budgetAllZones).

Yes, that is a more attractive approach. The reason I avoided it was because it wasn't clear how the config should refer to project level settings... i.e. fields not provided by mintr. But I will have a think.

> I ended up with a couple of duplicate strategies, is that supposed to be possible? It's because that's the best strategy at both those percentages? Could be confusing to the user? Not sure what best way of indicating it would be.
> Screenshot from 2021-07-22 11-03-48

Yes... in fact if you have a big budget then all the strategies will be the same. I think that for now this is best explained by the help material. Will fix.

> Edge case, but I think I get this error if I've created a region but not changed any settings for it yet:
> Screenshot from 2021-07-22 11-53-05
> 
> It did mean I got to test the nice dismissible errors!

This is indeed a problem - well spotted. I thought I had avoided it but clearly not. Will fix.

**Comment by EmmaLRussell** [comment]
> I meant the budget input form itself - it currently looks a bit like a dynamic form but it isn't!

:+1:

> Yes, that is a more attractive approach. The reason I avoided it was because it wasn't clear how the config should refer to project level settings... i.e. fields not provided by mintr. But I will have a think.
> 

Yeah, it could just assume that mintr will add them, which is a bit ugly.. So, instead, if projectOptions was a new endpoint, returning a dynamic form config which could be displayed in the strategize page.. and if the graphing code was updated to look in projectSettings for values as well as baseline and intevention.. I guess that would reduce the special knowledge code about that setting down to using it to fetch the strategies?

File: src/app/static/src/app/components/strategisePage.vue (hunk 1 above)
Comment (mwoodbri) [reply src/app/static/src/app/components/strategisePage.vue]: Done in 38def7e

File: src/app/static/src/app/components/figures/strategise/strategiesTable.vue (hunk 3 above)
Comment (mwoodbri) [reply src/app/static/src/app/components/figures/strategise/strategiesTable.vue:20]: 7899934

File: src/app/static/src/app/components/figures/strategise/strategiesTable.vue (hunk 4 above)
Comment (mwoodbri) [reply src/app/static/src/app/components/figures/strategise/strategiesTable.vue:13]: f60b175

**Comment by mwoodbri** [comment]
> > Edge case, but I think I get this error if I've created a region but not changed any settings for it yet:
> > Screenshot from 2021-07-22 11-53-05
> > It did mean I got to test the nice dismissible errors!
> 
> This is indeed a problem - well spotted. I thought I had avoided it but clearly not. Will fix.

Should be fixed by d7f885c. The code now ensures that only fully-initialised regions are included, based on checking for `budgetAllZones` (the field populated by the automatic form submission that takes place when the parameter form is loaded). This will be replaced by project settings in mrc-2533.

**Review by EmmaLRussell — APPROVED** [review]
Works great, bug has vanished, and I see the Strategize link appearing when I initialise a second region, as expected. 

> The code now ensures that only fully-initialised regions are included

This is  good solution, but we should probably make sure it gets mentioned in the user guide, so as to avoid confusion.

**Comment by mwoodbri** [comment]
> > The code now ensures that only fully-initialised regions are included
> 
> This is good solution, but we should probably make sure it gets mentioned in the user guide, so as to avoid confusion.

Thanks - I have a growing list, including explanations for why exclusively no-intervention strategies can be returned, and why strategies can be listed more than once in the table.

File: src/app/static/src/app/actions.ts [hunk 6]
```diff
@@ -169,6 +173,35 @@ export const actions: ActionTree<RootState, RootState> = {
...
             context.dispatch(RootAction.FetchCostTableConfig)
         ]);
+    },
+
+    [RootAction.DismissErrors](context) {
+        const {commit} = context;
+        commit(RootMutation.DismissErrors);
+    },
+
+    [RootAction.SetBudget](context, payload: { budget: number }) {
```
[the remark this answers is not in the document]
Comment (mwoodbri) [reply src/app/static/src/app/actions.ts:183]: Good point. I don't know why I thought that mutations should only be made from actions - especially as it's already done elsewhere! But I'll leave it as it is for now - at least until it's clear how the `projectOptions` stuff should work, just in case the action needs to do anything else.
